sdks/python/apache_beam/runners/portability/fn_api_runner.py

```diff
@@ -1319,55 +1320,62 @@ def stop_worker(self):
 @WorkerHandler.register_environment(common_urns.environments.DOCKER.urn,
                                     beam_runner_api_pb2.DockerPayload)
 class DockerSdkWorkerHandler(GrpcWorkerHandler):
+
+  _lock = threading.Lock()
```

Comment:
Add some comments as to why this is needed. Is it to workaround subprocess issues with Python 2? If so, perhaps lift it up as an appropriately named module global (as it should be shared with the SubProcess worker handler below).

Reply:
Yes, this is a workaround for subprocess.

I tried to add a function to handle calls to subprocess, but, I was not able to find a good way to handle it. With `DockerSdkWorkerHandler`, we can lock `start_worker()`, but we cannot lock `start_worker()` with `SubprocessSdkWorkerHandler` handler. We should lock `subprocess.Popen()` only with `SubprocessSdkWorkerHandler`, if we lock the whole `start_worker()` function, it waits a worker to finish job, so the pipeline gets stuck.

In fact, I doubt  `DockerSdkWorkerHandler` works with FnApi at the moment. FnApi doesn't stage any artifacts, so docker bootstrap fails and container cannot be created. I will make a change to `DockerSdkWorkerHandler` soon and create only one container and handle multithreading/processing when a job running within container, so we can get rid of lock here for `DockerSdkWorkerHandler`.

Reply:
> Yes, this is a workaround for subprocess.

Comments would be good for future readers of this code. 

> I tried to add a function to handle calls to subprocess, but, I was not able to find a good way to handle it. With `DockerSdkWorkerHandler`, we can lock `start_worker()`, but we cannot lock `start_worker()` with `SubprocessSdkWorkerHandler` handler. We should lock `subprocess.Popen()` only with `SubprocessSdkWorkerHandler`, if we lock the whole `start_worker()` function, it waits a worker to finish job, so the pipeline gets stuck.

I meant moving the lock out of DockerSdkWorkerHandler and SubprocessSdkWorkerHandler to a common place, e.g. a global _subprocess_lock. 

> In fact, I doubt `DockerSdkWorkerHandler` works with FnApi at the moment. FnApi doesn't stage any artifacts, so docker bootstrap fails and container cannot be created. 

It does work; this is how we run Java pipelines on the FnApiRunner. 

> I will make a change to `DockerSdkWorkerHandler` soon and create only one container and handle multithreading/processing when a job running within container, 

For all SDKs? 

> so we can get rid of lock here for `DockerSdkWorkerHandler`.

The lock is a log against all other calls to subprocess.